### PR TITLE
Fetch the mandatory claims from the JWT authenticator policy

### PR DESCRIPTION
**Pull request by nessiLahav**
### What does this PR do?
Fetch mandatory claims from JWT authenticator policy

#### Change log
- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [ ] The changes in this PR do not affect the Conjur API

### CODE REVIEW

File: app/domain/logs.rb [hunk 1]
```diff
@@ -614,6 +614,21 @@ module AuthnJwt
         code: "CONJ00121D"
       )
 
+      FetchingMandatoryClaims = ::Util::TrackableLogMessageClass.new(
```
Comment (nessiLahav) [inline app/domain/logs.rb:617]: @hilagross @shulifink Please review

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        },
+        inputs: %i[authentication_parameters]
+      ) do
+        def call
+          fetch_mandatory_claims
+        end
+
+        private
+
+        def fetch_mandatory_claims
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb:23]: Create function called mandatory claims with memoization

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        dependencies: {
+          resource_class: ::Resource,
+          fetch_required_secrets: ::Conjur::FetchRequiredSecrets.new,
+          parse_mandatory_claims: ::Authentication::AuthnJwt::InputValidation::ParseMandatoryClaims.new,
+          logger: Rails.logger
+        },
+        inputs: %i[authentication_parameters]
+      ) do
+        def call
+          fetch_mandatory_claims
```
Comment (benoaviram) [inline app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb:18]: Why do we need this function and not implement it here? do we have any future functions that will be called in this command class?

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        end
+
+        private
+
+        def fetch_mandatory_claims
+          @logger.debug(LogMessages::Authentication::AuthnJwt::FetchingMandatoryClaims.new)
+          if !mandatory_claims_resource_exists?
+            @logger.debug(LogMessages::Authentication::AuthnJwt::NotConfiguredMandatoryClaims.new)
+            return empty_mandatory_claims
+          else
```
Comment (benoaviram) [inline app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb:28]: what is the purpose of "else"? we will exit the function with the "return" when the if condition isn't mapped..

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb (hunk 3 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb:18]: This is our convention

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            fetch_mandatory_claims_secret_value
+            parse_mandatory_claims_secret_value
+          end
+
+          @logger.info(LogMessages::Authentication::AuthnJwt::FetchingMandatoryClaims.new(parse_mandatory_claims_secret_value))
```
Comment (benoaviram) [inline app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb:33]: it should be the FetchedMandatoryClaims right?

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+        def mandatory_claims_resource_id
+          @mandatory_claims_resource_id ||= "#{@authentication_parameters.authn_jwt_variable_id_prefix}/#{MANDATORY_CLAIMS_RESOURCE_NAME}"
+        end
+        
+        def empty_mandatory_claims
+          @empty_mandatory_claims ||= []
+        end
+        
+        def fetch_mandatory_claims_secret_value
```
Comment (benoaviram) [inline app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb:55]: What is the purpose of naming it secret? in the future we could use other way to save it, I dont think we should make the variable permanent

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb (hunk 3 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb]: I change it please review

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb (hunk 2 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb]: I change it please review

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb (hunk 4 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb]: fixed

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb (hunk 5 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb]: Nice catch

File: app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb (hunk 6 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/restriction_validators/fetch_mandatory_claims.rb:47]: Correct, but we have the dependency `FetchRequiredSecrets` to fetch the secret object and then i use it to get the value so i called it ...._secret_value

**Review by tzheleznyak — APPROVED** [review]
LGTM

**Review by benoaviram — APPROVED** [review]
LGTM

File: app/domain/logs.rb [hunk 7]
```diff
@@ -614,6 +614,21 @@ module AuthnJwt
...
+        msg: "Fetching mandatory claims...",
+        code: "CONJ00122D"
+      )
+
+      NotConfiguredMandatoryClaims = ::Util::TrackableLogMessageClass.new(
+        msg: "Mandatory claims is not configured",
```
Comment (hilagross) [inline app/domain/logs.rb:623]: I think mandatory claims **are** not configured
unless you are saying Mandatory claims **variable** is not configured

**Pull request merged**